File: quota/engine.py
"""Quota rules engine. Section 2 of the spec wins over the overview."""

from __future__ import annotations

import itertools
import random
from dataclasses import dataclass, field
from typing import Literal

from quota.cards import Card, bonus, make_deck, score_for

EndReason = Literal["DECK", "STALL"]


@dataclass(frozen=True, slots=True)
class TakeQuota:
    card_id: int

    def key(self) -> tuple[str, int]:
        return ("take", self.card_id)


@dataclass(frozen=True, slots=True)
class Collect:
    card_ids: tuple[int, ...]

    def key(self) -> tuple[str, tuple[int, ...]]:
        return ("collect", self.card_ids)


@dataclass(frozen=True, slots=True)
class Abandon:
    def key(self) -> tuple[str, int]:
        return ("abandon", -1)


@dataclass(frozen=True, slots=True)
class Pass:
    def key(self) -> tuple[str, int]:
        return ("pass", -1)


Action = TakeQuota | Collect | Abandon | Pass


@dataclass
class Player:
    name: str
    quota: Card | None = None
    collection: list[Card] = field(default_factory=list)
    achieved: list[Card] = field(default_factory=list)
    score: int = 0
    achieve_count: int = 0
    max_single_score: int = 0
    is_human: bool = True


@dataclass
class GameConfig:
    num_players: int = 3
    num_decks: int = 2
    market_size: int | None = None
    removed_count: int = 8
    stall_threshold: int | None = None
    stall_end_count: int = 2
    seed: int | None = None
    names: list[str] | None = None
    human_seats: list[int] | None = None

    def resolved_market_size(self) -> int:
        if self.market_size is not None:
            return self.market_size
        if self.num_players == 4:
            return 6
        return 7

    def resolved_stall_threshold(self) -> int:
        if self.stall_threshold is not None:
            return self.stall_threshold
        return self.num_players


@dataclass
class Game:
    config: GameConfig
    deck: list[Card]
    removed: list[Card]
    market: list[Card]
    discard: list[Card]
    players: list[Player]
    current: int
    no_gain_streak: int
    stall_flag: bool
    finished: bool
    end_reason: EndReason | None
    turn_number: int
    log: list[str]
    rng: random.Random
    reshuffle_count: int = 0

    @classmethod
    def start(cls, config: GameConfig | None = None) -> Game:
        cfg = config or GameConfig()
        if cfg.num_players < 2:
            raise ValueError("num_players must be at least 2")
        rng = random.Random(cfg.seed)
        deck = make_deck(cfg.num_decks)
        rng.shuffle(deck)
        removed_count = cfg.removed_count
        market_size = cfg.resolved_market_size()
        if len(deck) < removed_count + market_size:
            raise ValueError("deck is smaller than removed cards plus market")
        removed = [deck.pop() for _ in range(removed_count)]
        market = [deck.pop() for _ in range(market_size)]
        names = cfg.names or [f"P{i + 1}" for i in range(cfg.num_players)]
        if len(names) != cfg.num_players:
            raise ValueError("names length must match num_players")
        human = set(cfg.human_seats if cfg.human_seats is not None else range(cfg.num_players))
        players = [
            Player(name=names[i], is_human=i in human) for i in range(cfg.num_players)
        ]
        first = rng.randrange(cfg.num_players)
        game = cls(
            config=cfg,
            deck=deck,
            removed=removed,
            market=market,
            discard=[],
            players=players,
            current=first,
            no_gain_streak=0,
            stall_flag=False,
            finished=False,
            end_reason=None,
            turn_number=1,
            log=[],
            rng=rng,
        )
        game.log.append(f"先手: {players[first].name}")
        return game

    def market_size(self) -> int:
        return self.config.resolved_market_size()

    def legal_actions(self, seat: int | None = None) -> list[Action]:
        p = self.players[self.current if seat is None else seat]
        if p.quota is None:
            acts: list[Action] = [
                TakeQuota(c.id) for c in self.market if c.suit != "JOKER"
            ]
            return acts + [Pass()]
        assert p.quota.rank is not None
        eligible = [
            c.id
            for c in self.market
            if c.suit == p.quota.suit or c.suit == "JOKER"
        ]
        need = p.quota.rank - 1 - len(p.collection)
        acts: list[Action] = [
            Collect(ids)
            for size in range(1, need + 1)
            for ids in itertools.combinations(eligible, size)
        ]
        return acts + [Abandon(), Pass()]

    def step(self, action: Action) -> None:
        if self.finished:
            raise RuntimeError("game is already finished")
        action = self._canonicalize(action)
        legal = {a.key(): a for a in self.legal_actions()}
        if action.key() not in legal:
            raise ValueError(f"illegal action: {action}")
        p = self.players[self.current]
        gained = False
        if isinstance(action, TakeQuota):
            card = self._take_market(action.card_id)
            gained = True
            if card.rank == 1:
                self._achieve(p, [card], 1)
                self.log.append(f"{p.name} が {card.label()} の注文を請け負い、即納品（1点）")
            else:
                p.quota = card
                p.collection = []
                self.log.append(
                    f"{p.name} が {card.label()} の注文を請け負った（{card.rank}枚、{score_for(card.rank)}点）"
                )
        elif isinstance(action, Collect):
            assert p.quota is not None and p.quota.rank is not None
            taken = [self._take_market(card_id) for card_id in action.card_ids]
            gained = True
            p.collection.extend(taken)
            labels = "、".join(card.label() for card in taken)
            self.log.append(f"{p.name} が {labels} を買い付け")
            if 1 + len(p.collection) == p.quota.rank:
                rank = p.quota.rank
                cards = [p.quota, *p.collection]
                self._achieve(p, cards, rank)
                p.quota = None
                p.collection = []
                self.log.append(f"{p.name} が納品（{score_for(rank)}点）")
        elif isinstance(action, Abandon):
            assert p.quota is not None
            self.discard.extend([p.quota, *p.collection])
            self.log.append(f"{p.name} が注文を取り消した")
            p.quota = None
            p.collection = []
        elif isinstance(action, Pass):
            self.log.append(f"{p.name} はパス")
        else:
            raise TypeError(action)

        if gained:
            self.no_gain_streak = 0
            self.stall_flag = False
        else:
            self.no_gain_streak += 1
            if self.no_gain_streak == self.config.resolved_stall_threshold():
                if self.stall_flag:
                    self.finished = True
                    self.end_reason = "STALL"
                    self.log.append("交易の途絶（膠着の連続）")
                    return
                self.deck.extend(self.market)
                self.market = []
                self.rng.shuffle(self.deck)
                self.market = [self.deck.pop() for _ in range(self.market_size())]
                self.no_gain_streak = 0
                self.stall_flag = True
                self.reshuffle_count += 1
                self.log.append("新しい船団が入港した")

        self.current = (self.current + 1) % len(self.players)
        self.turn_number += 1
        self.begin_turn()

    def begin_turn(self) -> bool:
        """Refill the market at the start of a turn. False means the game ended."""
        while len(self.market) < self.market_size():
            if not self.deck:
                self.finished = True
                self.end_reason = "DECK"
                self.log.append("季節風の終わり（山札切れ）")
                return False
            self.market.append(self.deck.pop())
        return True

    def ranking(self) -> list[list[int]]:
        """Seats grouped best-first. Ties share a group."""
        seats = list(range(len(self.players)))
        seats.sort(
            key=lambda i: (
                self.players[i].score,
                self.players[i].achieve_count,
                self.players[i].max_single_score,
            ),
            reverse=True,
        )
        groups: list[list[int]] = []
        for seat in seats:
            if not groups or not self._tied(groups[-1][0], seat):
                groups.append([seat])
            else:
                groups[-1].append(seat)
        return groups

    def public_view(self) -> dict:
        """Observation without deck order or removed cards."""
        return {
            "market": [c.label() for c in self.market],
            "discard": [c.label() for c in self.discard],
            "deck_count": len(self.deck),
            "current": self.current,
            "no_gain_streak": self.no_gain_streak,
            "stall_flag": self.stall_flag,
            "finished": self.finished,
            "end_reason": self.end_reason,
            "turn_number": self.turn_number,
            "players": [
                {
                    "name": p.name,
                    "quota": None if p.quota is None else p.quota.label(),
                    "collection": [c.label() for c in p.collection],
                    "score": p.score,
                    "achieve_count": p.achieve_count,
                    "max_single_score": p.max_single_score,
                }
                for p in self.players
            ],
        }

    def all_card_ids(self) -> list[int]:
        ids: list[int] = []
        ids.extend(c.id for c in self.deck)
        ids.extend(c.id for c in self.removed)
        ids.extend(c.id for c in self.market)
        ids.extend(c.id for c in self.discard)
        for p in self.players:
            if p.quota is not None:
                ids.append(p.quota.id)
            ids.extend(c.id for c in p.collection)
            ids.extend(c.id for c in p.achieved)
        return ids

    def _canonicalize(self, action: Action) -> Action:
        if not isinstance(action, Collect):
            return action
        order = {card.id: index for index, card in enumerate(self.market)}
        return Collect(tuple(sorted(action.card_ids, key=lambda card_id: order.get(card_id, 10**9))))

    def _take_market(self, card_id: int) -> Card:
        for i, card in enumerate(self.market):
            if card.id == card_id:
                return self.market.pop(i)
        raise ValueError(f"card {card_id} is not in the market")

    def _achieve(self, player: Player, cards: list[Card], rank: int) -> None:
        player.achieved.extend(cards)
        pts = len(cards) + bonus(rank)
        player.score += pts
        player.achieve_count += 1
        player.max_single_score = max(player.max_single_score, pts)

    def _tied(self, a: int, b: int) -> bool:
        pa, pb = self.players[a], self.players[b]
        return (
            pa.score == pb.score
            and pa.achieve_count == pb.achieve_count
            and pa.max_single_score == pb.max_single_score
        )

File: unity/Assets/Scripts/Quota/Engine.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Quota
{
    public abstract class GameAction
    {
        public abstract string Key { get; }
    }

    public sealed class TakeQuota : GameAction
    {
        public readonly int CardId;
        public TakeQuota(int cardId) { CardId = cardId; }
        public override string Key => "take:" + CardId;
    }

    public sealed class Collect : GameAction
    {
        public readonly int[] CardIds;
        public Collect(IEnumerable<int> cardIds) { CardIds = cardIds.ToArray(); }
        public override string Key => "collect:" + string.Join(",", CardIds);
    }

    public sealed class Abandon : GameAction
    {
        public override string Key => "abandon";
    }

    public sealed class Pass : GameAction
    {
        public override string Key => "pass";
    }

    public sealed class Bundle
    {
        public readonly string Kind;
        public readonly bool HasWild;
        public Bundle(string kind, bool hasWild)
        {
            Kind = kind;
            HasWild = hasWild;
        }
    }

    public sealed class Player
    {
        public readonly string Name;
        public Card Quota;
        public readonly List<Card> Collection = new List<Card>();
        public readonly List<Card> Achieved = new List<Card>();
        public int Score;
        public int AchieveCount;
        public int MaxSingleScore;
        public readonly bool IsHuman;
        public readonly List<Bundle> Bundles = new List<Bundle>();
        public int ReshuffleTakeLeft;
        public int DoubleActionLeft;

        public Player(string name, bool isHuman)
        {
            Name = name;
            IsHuman = isHuman;
        }
    }

    public sealed class GameConfig
    {
        public int NumPlayers = 3;
        public int NumDecks = 2;
        public int? MarketSize;
        public int RemovedCount = 8;
        public int? StallThreshold;
        public int StallEndCount = 2;
        public int? Seed;
        public List<string> Names;
        public List<int> HumanSeats;
        public bool SequenceRule;
        public bool TitleRule;
        public int TitleMinAchieves = 3;
        public int TitleMonoBonus = 15;
        public int TitlePuristBonus = 5;
        public bool SpecialActionsRule;
        public int ReshuffleTakeUses = 1;
        public int DoubleActionUses = 1;
        public string ItemSet = "trade";

        public int ResolvedMarketSize()
        {
            if (MarketSize != null) return MarketSize.Value;
            return NumPlayers == 4 ? 6 : 7;
        }

        public int ResolvedStallThreshold()
        {
            if (StallThreshold != null) return StallThreshold.Value;
            return NumPlayers;
        }
    }

    public sealed class Game
    {
        public readonly GameConfig Config;
        public readonly List<Card> Deck;
        public readonly List<Card> Removed;
        public readonly List<Card> Market;
        public readonly List<Card> Discard = new List<Card>();
        public readonly List<Player> Players;
        public int Current;
        public int NoGainStreak;
        public bool StallFlag;
        public bool Finished;
        public string EndReason;
        public int TurnNumber;
        public readonly List<string> Log = new List<string>();
        public readonly PythonRandom Rng;
        public int ReshuffleCount;
        public bool TurnGain;
        public string Plan = "normal";
        public int DoubleStage;
        public bool DoubleGained;

        Game(GameConfig config, List<Card> deck, List<Card> removed, List<Card> market, List<Player> players, int current, PythonRandom rng)
        {
            Config = config;
            Deck = deck;
            Removed = removed;
            Market = market;
            Players = players;
            Current = current;
            Rng = rng;
            TurnNumber = 1;
        }

        public static Game Start(GameConfig config = null)
        {
            var cfg = config ?? new GameConfig();
            if (cfg.NumPlayers < 2) throw new ArgumentException("num_players must be at least 2");
            PythonRandom rng;
            if (cfg.Seed == null) rng = new PythonRandom(Environment.TickCount);
            else rng = new PythonRandom(cfg.Seed.Value);
            var deck = Cards.MakeDeck(cfg.NumDecks);
            rng.Shuffle(deck);
            var removedCount = cfg.RemovedCount;
            var marketSize = cfg.ResolvedMarketSize();
            if (deck.Count < removedCount + marketSize)
                throw new ArgumentException("deck is smaller than removed cards plus market");
            var removed = PopMany(deck, removedCount);
            var market = PopMany(deck, marketSize);
            var names = cfg.Names ?? new List<string>();
            if (cfg.Names == null)
            {
                for (var i = 0; i < cfg.NumPlayers; i++) names.Add($"P{i + 1}");
            }
            if (names.Count != cfg.NumPlayers) throw new ArgumentException("names length must match num_players");
            var human = new HashSet<int>(cfg.HumanSeats ?? Enumerable.Range(0, cfg.NumPlayers));
            var takeLeft = cfg.SpecialActionsRule ? cfg.ReshuffleTakeUses : 0;
            var doubleLeft = cfg.SpecialActionsRule ? cfg.DoubleActionUses : 0;
            var players = new List<Player>();
            for (var i = 0; i < cfg.NumPlayers; i++)
            {
                var player = new Player(names[i], human.Contains(i));
                player.ReshuffleTakeLeft = takeLeft;
                player.DoubleActionLeft = doubleLeft;
                players.Add(player);
            }
            var first = rng.RandBelow(cfg.NumPlayers);
            var game = new Game(cfg, deck, removed, market, players, first, rng);
            game.Log.Add($"先手: {players[first].Name}");
            return game;
        }

        public int MarketSize()
        {
            return Config.ResolvedMarketSize();
        }

        public List<GameAction> LegalActions(int? seat = null)
        {
            var player = Players[seat ?? Current];
            var actions = new List<GameAction>();
            if (player.Quota == null)
            {
                foreach (var card in Market)
                    if (card.Suit != Suit.Joker) actions.Add(new TakeQuota(card.Id));
                actions.Add(new Pass());
                return actions;
            }
            var eligible = new List<int>();
            foreach (var card in Market)
                if (card.Suit == player.Quota.Suit || card.Suit == Suit.Joker) eligible.Add(card.Id);
            var need = player.Quota.Rank.Value - 1 - player.Collection.Count;
            for (var size = 1; size <= need; size++)
            {
                foreach (var ids in Combinations(eligible, size))
                    actions.Add(new Collect(ids));
            }
            actions.Add(new Abandon());
            actions.Add(new Pass());
            return actions;
        }

        public void Step(GameAction action)
        {
            if (Finished) throw new InvalidOperationException("game is already finished");
            if (!IsLegal(action)) throw new ArgumentException($"illegal action: {action}");
            var player = Players[Current];
            var gained = false;
            if (action is TakeQuota take)
            {
                var card = TakeMarket(take.CardId);
                gained = true;
                if (card.Rank == 1)
                {
                    Achieve(player, new List<Card> { card }, 1);
                    Log.Add($"{player.Name} が {card.Label()} をノルマ札にし、即達成（1点）");
                }
                else
                {
                    player.Quota = card;
                    player.Collection.Clear();
                    Log.Add($"{player.Name} が {card.Label()} をノルマ札にした（{card.Rank}枚、{Cards.ScoreFor(card.Rank.Value)}点）");
                }
            }
            else if (action is Collect collect)
            {
                var taken = new List<Card>();
                foreach (var cardId in collect.CardIds) taken.Add(TakeMarket(cardId));
                gained = true;
                player.Collection.AddRange(taken);
                Log.Add($"{player.Name} が {string.Join("、", taken.ConvertAll(card => card.Label()))} を収集");
                TurnGain = true;
                if (1 + player.Collection.Count == player.Quota.Rank)
                {
                    var rank = player.Quota.Rank.Value;
                    var cards = new List<Card> { player.Quota };
                    cards.AddRange(player.Collection);
                    Achieve(player, cards, rank);
                    player.Quota = null;
                    player.Collection.Clear();
                    Log.Add($"{player.Name} がノルマ達成（{Cards.ScoreFor(rank)}点）");
                }
                else if (player.Collection.Count > 0 && !CanCollectMore(player))
                {
                    Log.Add($"{player.Name} は取れる札を取り切った");
                }
                else
                {
                    return;
                }
            }
            else if (action is Abandon)
            {
                Discard.Add(player.Quota);
                Discard.AddRange(player.Collection);
                Log.Add($"{player.Name} がノルマを放棄した");
                player.Quota = null;
                player.Collection.Clear();
            }
            else if (action is Pass)
            {
                Log.Add($"{player.Name} はパス");
            }
            else
            {
                throw new ArgumentException("unknown action");
            }

            CloseAction(gained);
        }

        public void DeclareReshuffle()
        {
            var player = Players[Current];
            Declare("reshuffle");
            SelfReshuffle();
            Log.Add($"{player.Name} が配り直し＆取得を宣言し、場を配り直した");
        }

        public void DeclareDouble()
        {
            var player = Players[Current];
            Declare("double");
            DoubleStage = 1;
            DoubleGained = false;
            Log.Add($"{player.Name} がダブルアクションを宣言した");
        }

        public void CancelDouble()
        {
            if (Finished) throw new InvalidOperationException("game is already finished");
            if (Plan != "double" || DoubleStage != 1 || TurnGain)
                throw new ArgumentException("ダブルアクションは取り消せません");
            var player = Players[Current];
            player.DoubleActionLeft += 1;
            Plan = "normal";
            DoubleStage = 0;
            DoubleGained = false;
            Log.Add($"{player.Name} がダブルアクションの宣言を取り消した");
        }

        void Declare(string kind)
        {
            if (Finished) throw new InvalidOperationException("game is already finished");
            if (!Config.SpecialActionsRule) throw new ArgumentException("特殊アクションは採用されていません");
            if (Plan != "normal" || TurnGain) throw new ArgumentException("この手番では特殊アクションを宣言できません");
            var player = Players[Current];
            if (kind == "reshuffle")
            {
                if (player.ReshuffleTakeLeft < 1) throw new ArgumentException("配り直し＆取得は使い切っています");
                player.ReshuffleTakeLeft -= 1;
                Plan = "reshuffle";
                return;
            }
            if (kind == "double")
            {
                if (player.DoubleActionLeft < 1) throw new ArgumentException("ダブルアクションは使い切っています");
                player.DoubleActionLeft -= 1;
                Plan = "double";
                return;
            }
            throw new ArgumentException(kind);
        }

        void SelfReshuffle()
        {
            Deck.AddRange(Market);
            Market.Clear();
            Rng.Shuffle(Deck);
            Market.AddRange(PopMany(Deck, MarketSize()));
        }

        void CloseAction(bool gained)
        {
            gained = gained || TurnGain;
            if (Plan == "double" && DoubleStage == 1)
            {
                DoubleGained = gained;
                TurnGain = false;
                if (!BeginTurn()) return;
                DoubleStage = 2;
                return;
            }
            EndTurn(gained || DoubleGained);
        }

        void EndTurn(bool gained)
        {
            if (gained)
            {
                NoGainStreak = 0;
                StallFlag = false;
            }
            else
            {
                NoGainStreak++;
                if (NoGainStreak == Config.ResolvedStallThreshold())
                {
                    if (StallFlag)
                    {
                        Finished = true;
                        EndReason = "STALL";
                        Log.Add("膠着の連続");
                        return;
                    }
                    SelfReshuffle();
                    NoGainStreak = 0;
                    StallFlag = true;
                    ReshuffleCount++;
                    Log.Add("場を配り直した");
                }
            }

            TurnGain = false;
            Plan = "normal";
            DoubleStage = 0;
            DoubleGained = false;
            Current = (Current + 1) % Players.Count;
            TurnNumber++;
            BeginTurn();
        }

        public bool BeginTurn()
        {
            while (Market.Count < MarketSize())
            {
                if (Deck.Count == 0)
                {
                    Finished = true;
                    EndReason = "DECK";
                    Log.Add("山札切れ");
                    return false;
                }
                Market.Add(Pop(Deck));
            }
            return true;
        }

        public List<List<int>> Ranking()
        {
            var seats = Enumerable.Range(0, Players.Count)
                .OrderByDescending(i => FinalScore(Players[i]))
                .ThenByDescending(i => Players[i].AchieveCount)
                .ThenByDescending(i => Players[i].MaxSingleScore)
                .ToList();
            var groups = new List<List<int>>();
            foreach (var seat in seats)
            {
                if (groups.Count == 0 || !Tied(groups[groups.Count - 1][0], seat))
                    groups.Add(new List<int>());
                groups[groups.Count - 1].Add(seat);
            }
            return groups;
        }

        public List<int> AllCardIds()
        {
            var ids = new List<int>();
            ids.AddRange(Deck.ConvertAll(card => card.Id));
            ids.AddRange(Removed.ConvertAll(card => card.Id));
            ids.AddRange(Market.ConvertAll(card => card.Id));
            ids.AddRange(Discard.ConvertAll(card => card.Id));
            foreach (var player in Players)
            {
                if (player.Quota != null) ids.Add(player.Quota.Id);
                ids.AddRange(player.Collection.ConvertAll(card => card.Id));
                ids.AddRange(player.Achieved.ConvertAll(card => card.Id));
            }
            return ids;
        }

        public int FinalScore(Player player)
        {
            return player.Score + SequencePoints(player) + TitlePoints(player);
        }

        public List<(string name, int points)> TitleAwards(Player player)
        {
            var awards = new List<(string, int)>();
            if (!Config.TitleRule) return awards;
            if (player.Bundles.Count < Config.TitleMinAchieves) return awards;
            var kinds = new HashSet<string>();
            var wild = false;
            foreach (var bundle in player.Bundles)
            {
                kinds.Add(bundle.Kind);
                if (bundle.HasWild) wild = true;
            }
            if (kinds.Count == 1) awards.Add(("単色達成", Config.TitleMonoBonus));
            if (!wild) awards.Add(("生粋の買い付け", Config.TitlePuristBonus));
            return awards;
        }

        public int TitlePoints(Player player)
        {
            var total = 0;
            foreach (var award in TitleAwards(player)) total += award.points;
            return total;
        }

        public int SequencePoints(Player player)
        {
            if (!Config.SequenceRule) return 0;
            return Cards.SequenceBonus(player.Achieved);
        }

        public bool IsLegal(GameAction action)
        {
            if (action is Collect collect)
            {
                var player = Players[Current];
                if (player.Quota == null || player.Quota.Rank == null) return false;
                var ids = collect.CardIds;
                if (ids.Length < 1 || ids.Distinct().Count() != ids.Length) return false;
                var need = player.Quota.Rank.Value - 1 - player.Collection.Count;
                if (ids.Length > need) return false;
                var eligible = new HashSet<int>();
                foreach (var card in Market)
                    if (card.Suit == player.Quota.Suit || card.Suit == Suit.Joker) eligible.Add(card.Id);
                foreach (var id in ids)
                    if (!eligible.Contains(id)) return false;
                return true;
            }
            var key = action.Key;
            foreach (var item in LegalActions())
                if (item.Key == key) return true;
            return false;
        }

        bool CanCollectMore(Player player)
        {
            if (player.Quota.Rank.Value - 1 - player.Collection.Count <= 0) return false;
            foreach (var card in Market)
                if (card.Suit == player.Quota.Suit || card.Suit == Suit.Joker) return true;
            return false;
        }

        Card TakeMarket(int cardId)
        {
            for (var i = 0; i < Market.Count; i++)
            {
                if (Market[i].Id != cardId) continue;
                var card = Market[i];
                Market.RemoveAt(i);
                return card;
            }
            throw new ArgumentException($"card {cardId} is not in the market");
        }

        void Achieve(Player player, List<Card> cards, int rank)
        {
            player.Achieved.AddRange(cards);
            var wild = false;
            foreach (var card in cards)
                if (card.Suit == Suit.Joker) wild = true;
            player.Bundles.Add(new Bundle(cards[0].Suit.ToString() == "Joker" ? "JOKER" : cards[0].Suit.ToString(), wild));
            var points = cards.Count + Cards.Bonus(rank);
            player.Score += points;
            player.AchieveCount++;
            if (points > player.MaxSingleScore) player.MaxSingleScore = points;
        }

        bool Tied(int a, int b)
        {
            var left = Players[a];
            var right = Players[b];
            return FinalScore(left) == FinalScore(right)
                && left.AchieveCount == right.AchieveCount
                && left.MaxSingleScore == right.MaxSingleScore;
        }

        static List<Card> PopMany(List<Card> deck, int count)
        {
            var taken = new List<Card>(count);
            for (var i = 0; i < count; i++) taken.Add(Pop(deck));
            return taken;
        }

        static Card Pop(List<Card> deck)
        {
            var card = deck[deck.Count - 1];
            deck.RemoveAt(deck.Count - 1);
            return card;
        }

        static IEnumerable<int[]> Combinations(IReadOnlyList<int> items, int size)
        {
            if (size <= 0 || size > items.Count) yield break;
            var index = new int[size];
            for (var i = 0; i < size; i++) index[i] = i;
            while (true)
            {
                var pick = new int[size];
                for (var i = 0; i < size; i++) pick[i] = items[index[i]];
                yield return pick;
                var cursor = size - 1;
                while (cursor >= 0 && index[cursor] == items.Count - size + cursor) cursor--;
                if (cursor < 0) yield break;
                index[cursor]++;
                for (var i = cursor + 1; i < size; i++) index[i] = index[i - 1] + 1;
            }
        }
    }
}
